### ISSUE TITLE
FSM for garbage

### ISSUE THREAD

**Issue by omenking** [omenking/swap-n-pop#166]
FSM will make it easier to write test code for garbage

**Comment by omenking** [comment]
For garbage I thought I'd try implement the change_state as the setter:

```ts
  set state(state) {
    if (this.state === state) { return; }
    this.state_timer = 0
    if (this.state_exit.has(this.state))
      this.state_exit.get(this.state)()
    this.state = state
    if (this.state_enter.has(this.state))
      this.state_enter.get(this.state)()
  }
```

There are cases where we set the state to null for garbage when its not in use. We might want to decide if garbage should always have a state and never be null. I don't know if the `null` state is the reason for my errors.

You can easily replicate the error by doing a 4x combo and the error chan be different but its generally a Maximum stack size succeeded

![screen shot 2017-12-12 at 4 09 33 pm](https://user-images.githubusercontent.com/7776/33908577-e2e0f3f6-df56-11e7-9f19-993a80b0b6db.png)

My thought is that the state is getting set repeatedly causing excessive calls. Garbage state probably gets set in panel which could be causing this loop. Or it could be relating to null state. or both.

**Comment by conormackeigan** [comment]
Stack overflow fixed in #168
